### PR TITLE
X.509 cert verifier

### PR DESCRIPTION
**Pull request by ripatel-fd**
- **secp256r1: add verify_no_low_s**
- **x509: better DER parser**
- **x509: X.509 parser, CA store and verify**
- **x509: check certificate validity periods**
- **x509: check key usage**
- **x509: add path length constraints**
- **x509: reject unknown critical extensions**

### CODE REVIEW

File: src/ballet/x509/fd_x509.c [hunk 1]
```diff
@@ -0,0 +1,779 @@
...
+        if( FD_UNLIKELY( fd_der_read_tl( &san, &gn_tag, &gn_len ) ) ) return -1;
+
+        (void)gn_tag;
+        san.p += gn_len;
+      }
+      continue;
+    }
+
+    /* Unknown extension */
+    if( FD_UNLIKELY( critical ) ) return -1;
```
[the remark this answers is not in the document]
Comment (ripatel-fd) [reply src/ballet/x509/fd_x509.c:342]: wont do

File: src/ballet/x509/fd_x509_verify.c [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    ulong ext_len = ( (ulong)p[0]<<8 ) | (ulong)p[1];
+    p += 2;
+    if( FD_UNLIKELY( (ulong)(list_end-p)<ext_len ) ) return FD_X509_VERIFY_ERR_PARSE;
+    uchar const * ext_end = p + ext_len;
+
+    while( p < ext_end ) {
+      if( FD_UNLIKELY( (ulong)(ext_end-p)<4UL ) ) return FD_X509_VERIFY_ERR_PARSE;
+      p += 2; /* ExtensionType */
+      ulong ext_data_len = ( (ulong)p[0]<<8 ) | (ulong)p[1];
+      p += 2;
```
[the remark this answers is not in the document]
Comment (ripatel-fd) [reply src/ballet/x509/fd_x509_verify.c:254]: doesnt matter

File: src/ballet/x509/fd_x509.c [hunk 3]
```diff
@@ -0,0 +1,829 @@
...
+      out->san_general_names_len = san_len;
+      out->has_subject_alt_name  = 1;
+
+      fd_der_cursor_t san = { .p = san_ptr, .end = san_ptr + san_len };
+      while( FD_DER_HAS_MORE( san ) ) {
+        int gn_tag; ulong gn_len;
+        if( FD_UNLIKELY( fd_der_read_tl( &san, &gn_tag, &gn_len ) ) ) return -1;
+
+        (void)gn_tag;
+        san.p += gn_len;
```
[the remark this answers is not in the document]
Comment (ripatel-fd) [reply src/ballet/x509/fd_x509.c:336]: Won’t do: the verifier only consumes dNSName, and unsupported GeneralName alternatives cannot affect hostname matching. The certificate signature authenticates the opaque bytes; a full GeneralName parser adds surface without changing authorization.

File: src/ballet/x509/fd_x509.c [hunk 4]
```diff
@@ -0,0 +1,829 @@
...
+          tbs.p += unique_id_len;
+          last_optional = field;
+        } else if( next_tag == (int)FD_DER_TAG_CONTEXT(3) ) {
+          if( FD_UNLIKELY( out->version!=2U || last_optional>=3 ) ) return -1;
+          last_optional = 3;
+          FD_DER_ENTER( tbs, FD_DER_TAG_CONTEXT(3) );
+            uchar const * ext_seq_ptr; ulong ext_seq_len;
+            FD_DER_READ( tbs, FD_DER_TAG_SEQUENCE, ext_seq_ptr, ext_seq_len );
+            fd_der_cursor_t ext = { .p = ext_seq_ptr, .end = ext_seq_ptr + ext_seq_len };
+            if( FD_UNLIKELY( fd_x509_parse_extensions( &ext, out ) ) ) return -1;
```
[the remark this answers is not in the document]
Comment (ripatel-fd) [reply src/ballet/x509/fd_x509.c:513]: Fixed with a direct SIZE (1..MAX) check; intentionally no dedicated edge-case test.

File: src/ballet/x509/fd_x509_ca_store.c [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    if( !begin ) break;
+    char const * b64 = begin + PEM_BEGIN_SZ;
+
+    char const * finish = find_line( b64, end, PEM_END, PEM_END_SZ );
+    char const * next_begin = find_line( b64, end, PEM_BEGIN, PEM_BEGIN_SZ );
+    if( next_begin && ( !finish || next_begin<finish ) ) {
+      p = next_begin;
+      continue;
+    }
+    if( !finish ) break;
```
[the remark this answers is not in the document]
Comment (ripatel-fd) [reply src/ballet/x509/fd_x509_ca_store.c]: Fixed: locate the next BEGIN first and bound the END search to it, making malformed marker scanning linear. Existing unterminated-block coverage exercises the behavior.

File: src/ballet/x509/fd_x509.c [hunk 6]
```diff
@@ -0,0 +1,829 @@
...
+      uchar const * serial; ulong serial_len;
+      FD_DER_READ( tbs, FD_DER_TAG_INTEGER, serial, serial_len );
+      if( FD_UNLIKELY( !fd_x509_serial_valid( serial, serial_len ) ) ) return -1;
+
+      /* signature AlgorithmIdentifier SEQUENCE */
+      FD_DER_READ( tbs, FD_DER_TAG_SEQUENCE, tbs_sig_alg, tbs_sig_alg_len );
+      out->sig_alg = fd_x509_parse_sig_alg( tbs_sig_alg, tbs_sig_alg_len );
+
+      /* issuer Name SEQUENCE */
+      FD_DER_READ_RAW( tbs, FD_DER_TAG_SEQUENCE, out->issuer, out->issuer_len );
```
[the remark this answers is not in the document]
Comment (ripatel-fd) [reply src/ballet/x509/fd_x509.c:467]: Won’t do: signature verification remains required for every chain link and anchor, so malformed-but-signed Name structure does not bypass trust. A full RDN/ATV parser is disproportionate here.

File: src/ballet/x509/fd_x509.c [hunk 7]
```diff
@@ -0,0 +1,829 @@
...
+      /* validity SEQUENCE { notBefore, notAfter } */
+      FD_DER_ENTER( tbs, FD_DER_TAG_SEQUENCE );
+        FD_DER_READ_TIME( tbs, out->not_before_tag, out->not_before, out->not_before_len );
+        FD_DER_READ_TIME( tbs, out->not_after_tag,  out->not_after,  out->not_after_len  );
+        out->not_before_unix = fd_x509_time_parse( out->not_before_tag, out->not_before, out->not_before_len );
+        out->not_after_unix  = fd_x509_time_parse( out->not_after_tag,  out->not_after,  out->not_after_len  );
+      FD_DER_LEAVE( tbs );
+
+      /* subject Name SEQUENCE */
+      FD_DER_READ_RAW( tbs, FD_DER_TAG_SEQUENCE, out->subject, out->subject_len );
```
[the remark this answers is not in the document]
Comment (ripatel-fd) [reply src/ballet/x509/fd_x509.c:478]: Won’t do: this is profile canonicality only. Authentication still requires a valid certificate signature and a matching subjectAltName, regardless of whether the SAN is marked critical; tracking extra parser state would not change authorization.
